core/src/internal/event_system.c

```diff
@@ -0,0 +1,333 @@
+					destroy_event_system(result);
+					result = NULL;
+				}
+			}
+		}
+	}
+	return result;
+}
+
+void EventSystem_AddEventCallback(EVENTSYSTEM_HANDLE event_system, GATEWAY_EVENT event_type, GATEWAY_CALLBACK callback)
+{
+	/* Codes_SRS_EVENTSYSTEM_26_012: [ This function shall log a failure and do nothing else when either `event_system` or `callback` parameters are NULL. ] */
+	if (event_system == NULL || callback == NULL)
+	{
+		LogError("null gateway events handle or callback parameters during event registration");
+	}
+	/* Codes_SRS_EVENTSYSTEM_26_011: [ This function shall register given GATEWAY_CALLBACK and call it when given GATEWAY_EVENT event happens inside of the gateway. ] */
+	else if (VECTOR_push_back(event_system->event_callbacks[event_type], &callback, 1) != 0)
+	{
+		/* Codes_SRS_EVENTSYSTEM_26_013: [ Should the worker thread ever fail to be created or any internall callbacks fail, failure will be logged and no further callbacks will be called during gateway's lifecycle. ] */
+		LogError("failed to register callback");
+		event_system->is_errored = 1;
+	}
+}
+
+void EventSystem_ReportEvent(EVENTSYSTEM_HANDLE event_system, GATEWAY_HANDLE gw, GATEWAY_EVENT event_type)
+{
+	/* Codes_SRS_EVENTSYSTEM_26_014: [ This function shall do nothing when `event_system` parameter is NULL. ] */
+	if (event_system == NULL)
+	{
+		LogError("null gateway handle or gateway event handle when reporting event");
+	}
+	/* Codes_SRS_EVENTSYSTEM_26_013: [ Should the worker thread ever fail to be created or any internall callbacks fail, failure will be logged and no further callbacks will be called during gateway's lifecycle. ] */
+	else if (!event_system->is_errored)
+	{
+		/* Lock-avoiding mechanism, we get a probably-past state with previous if, then check synchronized state to be sure */
+		int real_is_errored = 0;
+		Lock(event_system->internal_change_lock);
+		real_is_errored = event_system->is_errored;
+		Unlock(event_system->internal_change_lock);
+		
+		if (!real_is_errored)
+		{
+			/* We need to copy the callback queue because the callback might register another function */
+			/* Codes_SRS_EVENTSYSTEM_26_007: [ This function shan't call any callbacks registered for any other GATEWAY_EVENT other than the one given as parameter. ] */
+			VECTOR_HANDLE callbacks = event_system->event_callbacks[event_type];
+			size_t vector_size = VECTOR_size(callbacks);
+
+			if (vector_size > 0)
+			{
+				VECTOR_HANDLE call_queue = VECTOR_create(sizeof(GATEWAY_CALLBACK));
+				if (call_queue == NULL)
+				{
+					/*Codes_SRS_EVENTSYSTEM_26_013: [ Should the worker thread ever fail to be created or any internall callbacks fail, failure will be logged and no further callbacks will be called during gateway's lifecycle. ] */
+					LogError("Failed to create call queue during event report");
+					event_system->is_errored = 1;
+				}
+				else
+				{
+					if (VECTOR_push_back(call_queue, VECTOR_front(callbacks), vector_size) != 0)
```

Comment:
It looks like you're copying the callback pointers into a temporary vector to essentially take a snapshot of the registered listeners so that we can safely iterate through the vector without needing to lock each access into the vector. If that's the intent here then we should probably do this copy inside a lock.